### ISSUE TITLE
Missing ASSERT / bug in arithm.cpp

### ISSUE THREAD

**Issue by ArkadiuszRaj** [opencv/opencv#7792]
##### System information (version)
- OpenCV => 3.1 ( 380cc52)
- Operating System / Platform => Linux 64 Bit
- Compiler => gdb

##### Detailed description

Missing ASSERT /  bug in arithm.cpp

> size_t esz = src1.elemSize();
> size_t blocksize0 = (size_t)(BLOCK_SIZE + esz-1)/esz;

allows division by zero...

```
Thread 1 "pfnmc" received signal SIGFPE, Arithmetic exception.
0x00007ffff435146e in cv::compare (_src1=..., _src2=..., _dst=..., op=1) at /home/op/opencv/modules/core/src/arithm.cpp:1280
1280	    size_t blocksize0 = (size_t)(BLOCK_SIZE + esz-1)/esz;
(gdb) bt
#0  0x00007ffff435146e in cv::compare (_src1=..., _src2=..., _dst=..., op=1) at /home/op/opencv/modules/core/src/arithm.cpp:1280
#1  0x00007ffff44dd4fc in cv::MatOp_Cmp::assign (this=0x7ffff4fb8000 <cv::g_MatOp_Cmp>, e=..., m=..., _type=-1) at /home/op/opencv/modules/core/src/matop.cpp:1456
```

Excerpt from the code causing this behavior
```
Mat temp;
subtract(src(interior), mean[0], temp);
meanStdDev(temp, mean, stddev);
mask = temp > 5*mean[0];
```

**Comment by alalek** [comment]
`elemSize()` should be non-zero. Please provide complete minimal reproducer of this problem.

**Comment by ArkadiuszRaj** [comment]
That is complete code.

Before SIGPFE that is the state of the variables

*src* Mat content
```
[ 51,  50,  52,  51,  54,  58,  48,  53,  47,  66,  50,  46,  37,  44,  28,  20;
  52,  54,  48,  54,  50,  55,  56,  60,  47,  50,  50,  51,  43,  24,  28,  16;
  50,  55,  42,  49,  49,  54,  50,  56,  49,  37,  46,  39,  28,  22,  12,   7;
  53,  55,  53,  55,  54,  54,  50,  40,  48,  48,  45,  28,  21,  16,   8,   7]
```
*interior* rect definition (faulty, but...)
```
[12 x 0 from (2, 2)]
```
*temp* size() & elemSize()
```
[0 x 0] 0
```

**Comment by alalek** [comment]
Thank you for update!

I believe empty ROI is invalid usage for this expression: `src(interior)`. Same problem for empty Mat for arithmetic and other operations (probably we should raise "invalid argument" exception instead of silent logic problem).

/cc @vpisarev

BTW, There is responding code for `elemSize()`:
```
inline
size_t Mat::elemSize() const
{
    return dims > 0 ? step.p[dims - 1] : 0;
}
```

where `dims` is equal to `0` in your case (empty ROI -> empty image without any dimension/type information).

**Comment by ArkadiuszRaj** [comment]
Yeap, that is why I suggested ASSERT. The code snippet I have attached is simply an easy way for local "background" remover task. The thing is I  wrongly verified Mat dimensions (hence empty rect selector) - what happens on the image edge (cell rect trimmed to image rect). 

To sum up. Either assume that developer is keen to run gdb to hunt for the line and ignore this issue or (what I prefer) add an assert to explicitly point invalid argument.
